File: py/plugins/google-genai/src/genkit/plugins/google_genai/models/virtual_try_on.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Virtual Try-On image editing model for Google Vertex AI plugin.

Virtual Try-On takes a photo of a person and one or more photos of products
(garments) and returns an image of the person wearing the product. It is
available through Vertex AI only.

Callers identify which image is which by setting a ``type`` key in the
``ai.Part`` metadata — ``personImage`` for the model's photo and
``productImage`` for each garment. The string values match the Go and JS
conventions so the same request shape works across runtimes.
"""

import base64
import json
import sys
import urllib.parse
from typing import Any

if sys.version_info < (3, 11):
    from strenum import StrEnum
else:
    from enum import StrEnum

from google import genai
from pydantic import BaseModel, ConfigDict, Field

from genkit import (
    FinishReason,
    Media,
    MediaPart,
    Message,
    ModelInfo,
    ModelRequest,
    ModelResponse,
    Part,
    Role,
    Supports,
)
from genkit.plugin_api import ActionRunContext, tracer


class VirtualTryOnVersion(StrEnum):
    """Supported virtual try-on image editing models."""

    VIRTUAL_TRY_ON_001 = 'virtual-try-on-001'


# Virtual Try-On models available on the Vertex AI backend. Hardcoded because
# client.models.list() does not always surface them, and they must be visible
# in the plugin's Init / list_actions output to be selectable.
VERTEXAI_KNOWN_VIRTUAL_TRY_ON_MODELS: tuple[str, ...] = (VirtualTryOnVersion.VIRTUAL_TRY_ON_001,)


# Metadata keys used to tag input images. Users attach
# ``metadata={'type': PART_METADATA_TYPE_PERSON_IMAGE}`` to the person photo
# and ``{'type': PART_METADATA_TYPE_PRODUCT_IMAGE}`` to each garment.
PART_METADATA_TYPE_PERSON_IMAGE = 'personImage'
PART_METADATA_TYPE_PRODUCT_IMAGE = 'productImage'


def is_virtual_try_on_model(name: str) -> bool:
    """Check whether a model name denotes a virtual try-on model."""
    return name.startswith('virtual-try-on-')


class VirtualTryOnOutputOptions(BaseModel):
    """Output image format controls."""

    mime_type: str | None = Field(default=None, alias='mimeType')
    compression_quality: int | None = Field(default=None, alias='compressionQuality')

    model_config = ConfigDict(populate_by_name=True, extra='allow')


class VirtualTryOnConfig(BaseModel):
    """Configuration for a virtual-try-on request.

    Mirrors the JS ``ImagenTryOnConfigSchema`` in
    ``js/plugins/google-genai/src/vertexai/imagen.ts``.
    """

    sample_count: int | None = Field(default=None, ge=1, alias='sampleCount')
    seed: int | None = Field(default=None)
    base_steps: int | None = Field(default=None, ge=1, le=100, alias='baseSteps')
    person_generation: str | None = Field(default=None, alias='personGeneration')
    safety_setting: str | None = Field(default=None, alias='safetySetting')
    storage_uri: str | None = Field(default=None, alias='storageUri')
    output_options: VirtualTryOnOutputOptions | None = Field(default=None, alias='outputOptions')
    location: str | None = Field(default=None)

    model_config = ConfigDict(populate_by_name=True, extra='allow')


VIRTUAL_TRY_ON_SUPPORTS = Supports(
    media=True,
    multiturn=False,
    tools=False,
    system_role=True,
    output=['media'],
)


VIRTUAL_TRY_ON_MODEL_INFO = ModelInfo(
    label='Vertex AI - Virtual Try-On',
    supports=VIRTUAL_TRY_ON_SUPPORTS,
)


def virtual_try_on_model_info(version: str) -> ModelInfo:
    """ModelInfo for a virtual try-on model."""
    if version == VirtualTryOnVersion.VIRTUAL_TRY_ON_001:
        return ModelInfo(label='Vertex AI - Virtual Try-On 001', supports=VIRTUAL_TRY_ON_SUPPORTS)
    return ModelInfo(label=f'Vertex AI - {version}', supports=VIRTUAL_TRY_ON_SUPPORTS)


def _extract_media_by_type(request: ModelRequest, part_type: str) -> list[dict[str, Any]]:
    """Collect input images tagged with the given metadata type.

    Handles both data URIs (extracting base64 payload) and ``gs://`` URIs
    (passed through as ``gcsUri``).
    """
    out: list[dict[str, Any]] = []
    for message in request.messages:
        for part in message.content:
            root = part.root
            # Only media parts can carry image input
            if not isinstance(root, MediaPart):
                continue
            metadata = getattr(root, 'metadata', None) or {}
            if metadata.get('type') != part_type:
                continue
            url = root.media.url or ''
            if url.startswith('gs://'):
                out.append({'image': {'gcsUri': url}})
                continue
            # data:<mime>[;base64],<payload>
            if url.startswith('data:'):
                prefix, _, payload = url.partition(',')
                if not payload:
                    continue
                if ';base64' in prefix:
                    out.append({'image': {'bytesBase64Encoded': payload}})
                else:
                    # Non-base64 data URIs are rare; decode percent-encoding
                    # and re-encode as base64 for the Vertex API.
                    decoded = urllib.parse.unquote_to_bytes(payload)
                    out.append({'image': {'bytesBase64Encoded': base64.b64encode(decoded).decode('ascii')}})
    return out


def _to_virtual_try_on_request(request: ModelRequest, config: VirtualTryOnConfig | dict | None) -> dict[str, Any]:
    """Build the Vertex predict request body for virtual try-on."""
    persons = _extract_media_by_type(request, PART_METADATA_TYPE_PERSON_IMAGE)
    products = _extract_media_by_type(request, PART_METADATA_TYPE_PRODUCT_IMAGE)
    if not persons:
        raise ValueError(f"virtual try-on requires a media part with metadata.type='{PART_METADATA_TYPE_PERSON_IMAGE}'")
    if not products:
        raise ValueError(
            f"virtual try-on requires at least one media part with metadata.type='{PART_METADATA_TYPE_PRODUCT_IMAGE}'"
        )

    instance: dict[str, Any] = {
        'personImage': persons[0],
        'productImages': products,
    }

    parameters: dict[str, Any] = {}
    if isinstance(config, VirtualTryOnConfig):
        # The ``location`` field is a plugin-level override; it is not a Vertex
        # request parameter, so exclude it from the wire payload.
        parameters = config.model_dump(by_alias=True, exclude_none=True, exclude={'location'})
    elif isinstance(config, dict):
        parameters = {k: v for k, v in config.items() if k not in ('location', 'Location') and v is not None}

    return {
        'instances': [instance],
        'parameters': parameters,
    }


def _from_virtual_try_on_response(response: dict[str, Any]) -> list[Part]:
    """Convert a virtual-try-on predict response to MediaParts."""
    content: list[Part] = []
    for prediction in response.get('predictions') or []:
        b64 = prediction.get('bytesBase64Encoded') or ''
        mime_type = prediction.get('mimeType') or 'image/png'
        content.append(
            Part(
                root=MediaPart(
                    media=Media(
                        url=f'data:{mime_type};base64,{b64}',
                        content_type=mime_type,
                    )
                )
            )
        )
    return content


class VirtualTryOnModel:
    """Virtual-try-on model driver."""

    def __init__(self, version: str | VirtualTryOnVersion, client: genai.Client) -> None:
        """Initialize the virtual try-on model.

        Args:
            version: The virtual try-on model version.
            client: The configured Vertex AI genai.Client.
        """
        self._version = version
        self._client = client

    def _get_config(self, request: ModelRequest) -> VirtualTryOnConfig | dict | None:
        if request.config is None:
            return None
        if isinstance(request.config, VirtualTryOnConfig):
            return request.config
        if isinstance(request.config, dict):
            return VirtualTryOnConfig.model_validate(request.config)
        # Fall back to generic dict coercion for BaseModel instances
        if isinstance(request.config, BaseModel):
            return VirtualTryOnConfig.model_validate(request.config.model_dump())
        return None

    async def generate(self, request: ModelRequest, _: ActionRunContext) -> ModelResponse:
        """Handle a virtual-try-on generation request."""
        api_client = getattr(self._client, '_api_client', None)
        if api_client is None or not getattr(api_client, 'vertexai', False):
            raise ValueError('Virtual Try-On is only available through the Vertex AI backend')

        config = self._get_config(request)
        payload = _to_virtual_try_on_request(request, config)
        path = f'publishers/google/models/{self._version}:predict'

        with tracer.start_as_current_span('virtual_try_on_predict') as span:
            span.set_attribute(
                'genkit:input',
                json.dumps({'model': self._version, 'body': payload}, default=str),
            )
            http_response = await api_client.async_request('POST', path, payload)
            body = http_response.body if hasattr(http_response, 'body') else http_response
            response = json.loads(body) if isinstance(body, (bytes, bytearray, str)) else body
            span.set_attribute('genkit:output', json.dumps(response, default=str))

        if not response.get('predictions'):
            # Vertex returning zero predictions almost always means safety
            # filters blocked the output — surface it as a blocked response.
            return ModelResponse(
                message=Message(role=Role.MODEL, content=[]),
                finish_reason=FinishReason.BLOCKED,
                finish_message='virtual try-on: no predictions returned (likely content-filtered)',
            )

        return ModelResponse(
            message=Message(role=Role.MODEL, content=_from_virtual_try_on_response(response)),
        )

File: go/plugins/googlegenai/virtual_try_on.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package googlegenai

import (
	"context"
	"strings"

	"github.com/firebase/genkit/go/ai"
	"github.com/firebase/genkit/go/core/status"
	"github.com/firebase/genkit/go/plugins/internal/uri"
	"google.golang.org/genai"
)

// Part metadata keys used to tag images passed to virtual try-on.
//
// Callers should attach `Metadata: map[string]any{"type": "personImage"}` to
// the single media part representing the person, and
// `Metadata: map[string]any{"type": "productImage"}` to each media part
// representing a garment/product image.
const (
	PartMetadataTypePersonImage  = "personImage"
	PartMetadataTypeProductImage = "productImage"
)

// extractMediaByType walks the request messages and collects the media and data
// parts whose metadata["type"] matches typ. Uses the shared uri.Data helper
// (the same one veo.go uses) so the parsing of data:<mime>[;base64],<data>
// URIs is consistent across plugins. gs:// URIs are passed through as-is.
//
// A part tagged with typ that cannot be parsed is an error rather than a skip:
// dropping it would either report the image as missing or quietly send fewer
// product images than the caller asked for.
func extractMediaByType(input *ai.ModelRequest, typ string) ([]*genai.Image, error) {
	var out []*genai.Image
	for _, msg := range input.Messages {
		for _, p := range msg.Content {
			if !p.IsMedia() && !p.IsData() {
				continue
			}
			if metaType, _ := p.Metadata["type"].(string); metaType != typ {
				continue
			}
			switch {
			case strings.HasPrefix(p.Text, "gs://"):
				// uri.Data enforces this for the paths it handles; the fast
				// path here has to do it itself.
				if p.ContentType == "" {
					return nil, status.Errorf(status.ErrInvalidArgument, "virtual try-on: %s part from a gs:// URI must set a content type", typ)
				}
				out = append(out, &genai.Image{GCSURI: p.Text, MIMEType: p.ContentType})
			case strings.HasPrefix(p.Text, "http://"), strings.HasPrefix(p.Text, "https://"):
				// This branch takes no download middleware, and uri.Data would
				// hand back the URL's own bytes as if they were image data.
				return nil, status.Errorf(status.ErrInvalidArgument, "virtual try-on: %s part must be inline data or a gs:// URI; http(s) URLs are not fetched for this model", typ)
			default:
				mimeType, data, err := uri.Data(p)
				if err != nil {
					return nil, status.Errorf(status.ErrInvalidArgument, "virtual try-on: unreadable %s part: %w", typ, err)
				}
				out = append(out, &genai.Image{ImageBytes: data, MIMEType: mimeType})
			}
		}
	}
	return out, nil
}

// toRecontextImageSource builds the SDK's source from the tagged parts of the
// request: one person image and one or more product images.
func toRecontextImageSource(input *ai.ModelRequest) (*genai.RecontextImageSource, error) {
	persons, err := extractMediaByType(input, PartMetadataTypePersonImage)
	if err != nil {
		return nil, err
	}
	products, err := extractMediaByType(input, PartMetadataTypeProductImage)
	if err != nil {
		return nil, err
	}
	if len(persons) == 0 {
		return nil, status.Errorf(status.ErrInvalidArgument, "virtual try-on requires a media part with metadata.type=%q", PartMetadataTypePersonImage)
	}
	if len(products) == 0 {
		return nil, status.Errorf(status.ErrInvalidArgument, "virtual try-on requires at least one media part with metadata.type=%q", PartMetadataTypeProductImage)
	}

	if len(persons) > 1 {
		return nil, status.Errorf(status.ErrInvalidArgument, "virtual try-on accepts a single %s part, got %d", PartMetadataTypePersonImage, len(persons))
	}

	source := &genai.RecontextImageSource{PersonImage: persons[0]}
	for _, img := range products {
		source.ProductImages = append(source.ProductImages, &genai.ProductImage{ProductImage: img})
	}
	return source, nil
}

// generateVirtualTryOn requests a recontextualization call to the specified
// virtual try-on model with the provided configuration. The model dresses the
// person image in the product images; there is no text prompt.
func generateVirtualTryOn(
	ctx context.Context,
	client *genai.Client,
	model string,
	input *ai.ModelRequest,
	config *genai.RecontextImageConfig,
	cb func(context.Context, *ai.ModelResponseChunk) error,
) (*ai.ModelResponse, error) {
	if cb != nil {
		return nil, status.Errorf(status.ErrUnimplemented, "streaming mode not supported for virtual try-on")
	}
	// The SDK refuses this call on the Gemini Developer API, but says so in
	// terms of its own backend names; fail earlier with the plugin's.
	if client.ClientConfig().Backend != genai.BackendVertexAI {
		return nil, status.Errorf(status.ErrUnimplemented, "virtual try-on is only available through the Vertex AI backend")
	}

	source, err := toRecontextImageSource(input)
	if err != nil {
		return nil, err
	}

	resp, err := client.Models.RecontextImage(ctx, model, source, config)
	if err != nil {
		return nil, wrapAPIError(err)
	}

	// Vertex returning no images for a well-formed request almost always means
	// safety filters blocked the output. Surface that as a blocked response so
	// callers can handle it idiomatically, mirroring veo.go.
	if len(resp.GeneratedImages) == 0 {
		return &ai.ModelResponse{
			Message:       &ai.Message{Role: ai.RoleModel},
			FinishReason:  ai.FinishReasonBlocked,
			FinishMessage: "virtual try-on: no images returned (likely content-filtered)",
			Request:       input,
		}, nil
	}

	r := translateImagenCandidates(resp.GeneratedImages)
	r.Request = input
	return r, nil
}
